docs/design/2023-09-21-tiflash-resource-control.md

```diff
@@ -0,0 +1,73 @@
+
+##### Updating Resource Usage
+After TaskScheduler obtains a task, it will execute for 100ms and update its CPU usage status. During this time, it will call relevant interfaces of the LocalAdmissionController (LAC) to update token consumption.
+
+### LocalAdmissionController (LAC)
+LAC is responsible for managing the metadata of all resource groups on a TiFlash node, including:
+1. Recording the priorities, configurations, and TokenBuckets of all currently known resource groups.
+2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.
+3. Communicating with GAC to periodically obtain tokens.
+
+## Test Design
+### ResourceControlQueue Test
+1. Test when Resource Units (RU) are sufficient but physical resources are insufficient, the scheduling between different resource groups generally adheres to priority scheduling.
+2. Test when RU is insufficient but physical resources are sufficient, the proportion of physical resource usage between different resource groups is roughly equal to the `RU_PER_SEC` ratio.
+3. Test when RU and physical resources alternately become insufficient, the physical resource usage between different resource groups still meets expectations, meaning that the resource group with a higher `RU_PER_SEC` uses more physical resources.
+
+### LAC Test
+1. Create two resource groups (rg1, rg2), both with relatively large `RU_PER_SEC`, e.g., 3000, run a tpch workload, observe CPU usage, and query execution times. Expect that the query execution times of the two resource groups are roughly the same.
+2. Change the `RU_PER_SEC` of rg1 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg1.
+3. Change the `RU_PER_SEC` of rg2 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg2.
+4. Set the burstable attribute of rg1 to true, observe CPU usage, and query execution times. Expect increased CPU usage and significantly reduced query execution times for rg1.
+5. Test that LAC reports accurate RU usage to GAC.
+6. Test that the conversion between the three TokenBucket modes works as expected.
+7. Test with multiple TiFlash nodes, running the above tests (1/2/3) with tpch workload, expecting similar results.
+8. Test repeated deletion and reconstruction of resource groups. Expect queries to fail with "resource group not found" errors.
+
+## Impacts & Risks
+1. The introduction of ResourceControlQueue may lead to performance degradation as TaskScheduler's scheduling adds an extra layer of TaskQueue.
+2. TiFlash resource control only works when TiDB system variable `tidb_enable_resource_control` and TiFlash config `enable_resource_control` are both true. Other, flow control and priority scheduling will not take effects.
```

Comment:
```suggestion
2. TiFlash resource control only works when TiDB system variable `tidb_enable_resource_control` and TiFlash config `enable_resource_control` are both true. Otherwise, flow control and priority scheduling will not take effects.
```